sql/core/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala

```diff
@@ -81,11 +81,11 @@ case class UserDefinedFunction protected[sql] (
       f,
       dataType,
       exprs.map(_.expr),
+      nullableTypes.map(_.map(!_)).getOrElse(exprs.map(_ => false)),
```

Comment:
We need to be more conservative here. The default behavior needs to handle NULLs. We should fix it.

```
// Need to handle NULL when the fields can't accept NULL. 
val handleNullForInputs = nullableTypes.map(_.map(!_)).getOrElse(Seq.fill(exprs.size)(true))
```

Reply:
@srowen @cloud-fan This sounds a new behavior added in Spark 2.4 release by the PR https://github.com/apache/spark/pull/22259, right?

Reply:
It's new in the sense that Scala 2.12 forced us to change where we get type information for UDFS, or specifically whether the type is allowed to be null. Before we could infer it directly from the closure, but not in 2.12. This change is supposed to just get the info from elsewhere -- from TypeTags when the UDFs are declared. The change ought to be invisible to users and if done correctly should not result in a behavior change. This follow-up change is more about being conservative and lessening the chance of a mistake (see the test case mary ann found) internally or of a mistake for a library that uses ScalaUDF directly anyway. 

For that reason, yeah I agree with your comment here. Default to assuming null checks are needed.

Reply:
Looks like the only place where we'd get a not-specified `inputSchemas` is when `ScalaReflection.schemaFor` doesn't recognize a type and throws an exception (https://github.com/apache/spark/blob/1fd59c129a7aa16f9960b109128b166952992f32/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala#L786). The caller seems to be doing a bad job by calling it this way, for example:
```
    val inputSchemas = Try(ScalaReflection.schemaFor(typeTag[A1]) :: ScalaReflection.schemaFor(typeTag[A2]) :: Nil).toOption
    val udf = SparkUserDefinedFunction.create(f, dataType, inputSchemas)
```
It would mean if the type of only one of the parameters is unrecognizable by `ScalaReflection`, we'd end up having the entire `Seq` as `None`. I think it's fine not to check null for user-defined types that we don't know, coz they can't be primitive types anyway, but I do think we should make the type inference of each parameter independent so we do handle the nulls that need to be taken care of.

Reply:
Yes that's right. There are a number of UDFs in MLlib, etc that have inputs of type "Any", which isn't great, but I wanted to work around rather than change them for now.

Reply:
In addition to what I just pointed out, which is when we did try to get `inputSchemas` through `ScalaReflection.schemaFor` and got an exception for unrecognized types, there's another case where we could get an unspecified `nullableTypes`, and that is when `UserDefinedFunction` is instantiated calling the constructor but not the `create` method.
Then I assume it's created by an earlier version, and we should use the old logic, i.e., `ScalaReflection.getParameterTypes` (https://github.com/apache/spark/pull/22259/files#diff-57b3d87be744b7d79a9beacf8e5e5eb2L2153) to get the correct information for `nullableTypes`. Is that right, @cloud-fan @srowen ?

Reply:
Hm, but we can't use getParameterTypes anymore. It won't work in Scala 2.12. Where the nullability info is definitely not available, be conservative and assume it all needs null handling?